### PR TITLE
Reduce net consumption

### PR DESCRIPTION
**Pull request by alonSadan**
As an attempt to reduce the bridge-marker net consumption this PR increases the timeouts between the updates and uses a cache to keep track of the already reported bridges on the node. Only every ten iterations the info will be retrieved from the actual node object.
Also to avoid communication peeks a jitter is used to randomize the interval between updates.

```release-note
use cache to keep track of the reported bridges, instead of retrieving it from the API server
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,5 @@ hack/config-provider-*
 _out/
 
 _kubevirtci
+
+.idea/
```
Comment (phoracek) [inline .gitignore:12]: This is not related to this project. Please use a [global gitignore](https://stackoverflow.com/questions/7335420/global-git-ignore) to ignore files specific to your IDE

File: cmd/marker/main.go [hunk 2]
```diff
@@ -16,27 +16,38 @@ package main
 
 import (
 	"flag"
-	"time"
-
+	"github.com/Rican7/retry/jitter"
```
Comment (phoracek) [inline cmd/marker/main.go:19]: Let's use Jitter function from Kubernetes' API Machinery (which is already shipped in the project I believe) instead. https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#Jitter

File: cmd/marker/main.go [hunk 3]
```diff
@@ -16,27 +16,38 @@ package main
...
+			err := marker.Update(*nodeName, !useCache)
+			if err != nil {
+				glog.Errorf("Update failed: %v", err)
+			}
+		} else {
+			err := marker.Update(*nodeName, useCache)
+			if err != nil {
+				glog.Errorf("Update failed: %v", err)
+			}
+			// will use a default random generator
```
Comment (phoracek) [inline cmd/marker/main.go:48]: I think this comment is redundant. Although it points to a question - is this using crypto library for random numbers? If not, it may mean that all instances of the marker will get the same jittered numbers. Would you please check it?

File: cmd/marker/main.go [hunk 4]
```diff
@@ -16,27 +16,38 @@ package main
...
 	"github.com/golang/glog"
 	"github.com/kubevirt/bridge-marker/pkg/marker"
+	"time"
 )
 
 func main() {
 	nodeName := flag.String("node-name", "", "name of kubernetes node")
-	pollInterval := flag.Int("poll-interval", 10, "interval between updates in seconds, 10 by default")
+	updateInterval := flag.Int("update-interval", 60, "interval between updates in seconds, 60 by default")
+	pollInterval := flag.Int("poll-interval", 10, "interval between node info polls. every 10 cache updates by default")
```
Comment (phoracek) [inline cmd/marker/main.go:28]: Poll is not really describing this. Maybe "cacheResetInterval"?

File: pkg/marker/marker.go [hunk 5]
```diff
@@ -72,32 +73,38 @@ func getAvailableResources() (map[string]bool, error) {
...
-		Nodes().
-		Get(context.TODO(), nodeName, metav1.GetOptions{})
-	if err != nil {
-		return nil, fmt.Errorf("failed to get node: %v", err)
-	}
+func getReportedResources(nodeName string, useCache bool) (map[string]bool, error) {
+	if useCache {
+		return reportedResourcesCache, nil
+	} else {
+
```
Comment (phoracek) [inline pkg/marker/marker.go:80]: No need for this empty line

File: pkg/marker/marker.go [hunk 6]
```diff
@@ -110,6 +117,7 @@ func Update(nodeName string) error {
 				Op:   "remove",
 				Path: fmt.Sprintf("/status/capacity/%s~1%s", resourceNamespace, reportedResource),
 			})
+			delete(reportedResourcesCache, reportedResource)
```
Comment (phoracek) [inline pkg/marker/marker.go:120]: This is not correct. You are updating the cache before we get to apply the patch and get confirmation about its success.

**Review by phoracek — CHANGES_REQUESTED** [review]

File: cmd/marker/main.go [hunk 7]
```diff
@@ -24,7 +24,7 @@ import (
 
 func main() {
 	nodeName := flag.String("node-name", "", "name of kubernetes node")
-	pollInterval := flag.Int("poll-interval", 10, "interval between updates in seconds, 10 by default")
+	pollInterval := flag.Int("poll-interval", 60, "interval between updates in seconds, 60 by default")
```
Comment (qinqon) [inline cmd/marker/main.go]: Can we create a defaultPollInterval constant and use at b oth places ?

File: cmd/marker/main.go [hunk 8]
```diff
@@ -24,19 +24,28 @@ import (
...
+			err := marker.Update(*nodeName, !useCache)
+			if err != nil {
+				glog.Errorf("Update failed: %v", err)
+			}
+		} else {
+			err := marker.Update(*nodeName, useCache)
+			if err != nil {
+				glog.Errorf("Update failed: %v", err)
+			}
+			time.Sleep(time.Duration(*updateInterval) * time.Second)
```
Comment (qinqon) [inline cmd/marker/main.go]: Let's use the pollInmediate primitive here https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollImmediate

Also the fact the useCache mechanism has to be implemented inside Update, usually a timestamp is stored about the cache refresh time, then elapsed is calculated between refresah timestamp and current timestamps and if elapsed is >= pollInterval then we refresh the cache before doing anything anything and we continue.

**Review by alonSadan — COMMENTED** [review]
Thanks @phoracek  and @qinqon 
I added a cache package and managed to cache updates by the time passed from the last refresh, as Quique suggested.
Tell me what you think now

File: cmd/marker/main.go [hunk 9]
```diff
@@ -24,7 +25,8 @@ import (
 
 func main() {
 	nodeName := flag.String("node-name", "", "name of kubernetes node")
-	pollInterval := flag.Int("poll-interval", 10, "interval between updates in seconds, 10 by default")
+	const defaultPollInterval = 60
```
Comment (qinqon) [inline cmd/marker/main.go]: Let's use `time.Duration` from the beggining 

```golang
const defaultPollInterval = 60 * time.Second

pollInterval := flag.Int("update-interval", int(defaultPollInterval.Seconds()), fmt.Sprintf("interval between updates in seconds, %d by default", defaultPollInterval))

```

File: cmd/marker/main.go [hunk 10]
```diff
@@ -20,25 +20,39 @@ import (
...
 )
 
 func main() {
 	nodeName := flag.String("node-name", "", "name of kubernetes node")
-	const defaultPollInterval = 60
-	pollInterval := flag.Int("update-interval", defaultPollInterval, fmt.Sprintf("interval between updates in seconds, %d by default", defaultPollInterval))
+	const defaultUpdateInterval = 60
+	updateInterval := flag.Int("update-interval", defaultUpdateInterval, fmt.Sprintf("interval between updates in seconds, %d by default", defaultUpdateInterval))
+	const defaultReconcileInterval = 10
+	reconcileInterval := flag.Int("reconcile-interval", defaultReconcileInterval, fmt.Sprintf("interval between node bridges reconcile in minutes, %d by default", defaultReconcileInterval))
```
Comment (qinqon) [inline cmd/marker/main.go:33]: ditto, let's use duration for for intervals from the beggining so we can express quantity and unit at the same variable.

File: cmd/marker/main.go [hunk 11]
```diff
@@ -20,25 +20,39 @@ import (
...
 	}
 
-	for {
-		err := marker.Update(*nodeName)
+	cache := cache.Cache{}
+	wait.PollImmediateInfinite(time.Duration(*updateInterval) * time.Second, func() (bool, error) {
+		if time.Now().Sub(cache.LastRefreshTime()) >= time.Duration(*reconcileInterval) * time.Minute {
+			reportedBridges, err := marker.GetReportedResources(*nodeName)
+			if err != nil {
+				glog.Errorf("GetReportedResources failed: %v", err)
```
Comment (qinqon) [inline cmd/marker/main.go:47]: Don't we have to stop here ?

File: pkg/cache/cache.go [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+	c.bridges = freshBridges
+	c.lastRefreshTime = time.Now()
+}
+
+func (c *Cache) LastRefreshTime() time.Time {
+	return c.lastRefreshTime
+}
+
+func (c Cache) Bridges() map[string]bool {
+	return c.bridges
```
Comment (qinqon) [inline pkg/cache/cache.go:22]: We have to return a new map here to prevent cache being modified without using "Refresh"

You can check the possible problem here https://play.golang.org/p/-h9fqv2gdII

**Review by qinqon — CHANGES_REQUESTED** [review]

File: cmd/marker/main.go (hunk 11 above)
Comment (alonSadan) [reply cmd/marker/main.go:47]: I was not sure, maybe it is better to log the error and try again?

File: cmd/marker/main.go (hunk 10 above)
Comment (alonSadan) [reply cmd/marker/main.go:33]: Done

File: pkg/cache/cache.go (hunk 12 above)
Comment (alonSadan) [reply pkg/cache/cache.go:22]: nice catch!  and thanks for the playground explanation.
Done.

**Review by alonSadan — COMMENTED** [review]
Thanks Quique.
Addressed your comments

**Comment by qinqon** [comment]
/lgtm

We need @phoracek ack

File: cmd/marker/main.go [hunk 13]
```diff
@@ -16,27 +16,43 @@ package main
...
 	if *nodeName == "" {
 		glog.Fatal("node-name must be set")
 	}
 
-	for {
-		err := marker.Update(*nodeName)
+	cache := cache.Cache{}
+	wait.JitterUntil(func() {
+		jitteredReconcileInterval := wait.Jitter(time.Duration(*reconcileInterval)*time.Minute, 1.2)
+		if time.Now().Sub(cache.LastRefreshTime()) >= jitteredReconcileInterval {
```
Comment (phoracek) [inline cmd/marker/main.go:45]: Give this a name - assign this into a variable, so then you can just

```
if shouldReconcileNode
```

Or whatever makes sense here

File: pkg/marker/marker.go [hunk 14]
```diff
@@ -88,20 +89,15 @@ func getReportedResources(nodeName string) (map[string]bool, error) {
 			reportedResources[splitNodeResourceName[1]] = true
 		}
 	}
-
```
Comment (phoracek) [inline pkg/marker/marker.go:91]: Please keep the line

File: cmd/marker/main.go [hunk 15]
```diff
@@ -16,27 +16,43 @@ package main
...
+			reportedBridges, err := marker.GetReportedResources(*nodeName)
+			if err != nil {
+				glog.Errorf("GetReportedResources failed: %v", err)
+			}
+			cache.Refresh(reportedBridges)
```
Comment (phoracek) [inline cmd/marker/main.go:50]: We should add a log here to log a warning in case resources reported on the Node are not the same as kept in the cache. We don't expect it to happen, if it does, something may be wrong with the cluster and we want it to be visible in logs.

**Review by phoracek — CHANGES_REQUESTED** [review]
I like the abstraction of Cache. Some nits

File: cmd/marker/main.go (hunk 15 above)
Comment (alonSadan) [reply cmd/marker/main.go:57]: Done.

File: pkg/marker/marker.go (hunk 14 above)
Comment (alonSadan) [reply pkg/marker/marker.go]: Done.

File: cmd/marker/main.go (hunk 13 above)
Comment (alonSadan) [reply cmd/marker/main.go]: Done.

**Review by alonSadan — COMMENTED** [review]
Thanks Petr

File: .idea/workspace.xml [hunk 16]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (phoracek) [inline .idea/workspace.xml:1]: Please remove this (I would recommend adding .idea to global gitignore)

**Review by phoracek — CHANGES_REQUESTED** [review]

File: .idea/workspace.xml (hunk 16 above)
Comment (alonSadan) [reply .idea/workspace.xml]: oops 
and you already said it!
removed (and also added .idea to global gitignore)

**Review by phoracek — APPROVED** [review]
/lgtm
/approve
/hold

Looks great, thanks!

Holding in case @qinqon wants to take a final look on the minor changes

**Comment by qinqon** [comment]
/hold cancel

All good !
